src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1024,6 +1032,12 @@ private void ParseAndAddValue(HeaderDescriptor descriptor, HeaderStoreItemInfo i
         {
             Debug.Assert(info != null);
 
+            // We can get here only through Add methods which treat `value` as already parsed and do not add it into RawValue,
+            // so we have to manually add it into OriginalRawValue to return from non-validating enumeration.
+            object? originalRawValue = info.OriginalRawValue;
+            AddValueToStoreValue<string>(value ?? string.Empty, ref originalRawValue);
```

Comment:
This is going to add overhead on every header value set by the user. I think it would be better to avoid that if it's not needed.

Reply:
Could you please clarify how we can avoid it? We need to track all changes made by all `Add()` calls which do not changed `RawValue`.

Reply:
Just clear out the OriginalRawValue. That's what you are doing for set and remove, right? Why can't you do the same thing for Add?

Reply:
If we did that, we would lose the original raw value. In example, test [NonValidated_AddAndRemoveSingleValueHeader_EnumeratorReturnsNoValue](https://github.com/dotnet/runtime/blob/9c860a47d50cedbbe60f49ef891cd0b453bb0e9e/src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpHeadersTest.cs#L1785) would fail because instead of the expected "raw1", it would get "parsed1". So, I think we need to keep it around.

Reply:
Why do we care about keeping the original raw value in this case? If the user modifies the header through the API, then I think it's fine to discard the original raw value. In a lot of cases (like set and remove), we need to do discard it anyway, because it's not correct anymore. In this particular case, since they are adding the value as a string, I guess you could make the case that we can add another original raw value that corresponds to the value that they added. But it seems inconsistent, and I'm not sure what value it gets us. Seems better just to say, if you modify stuff via parsed header API (i.e. anything except TryAddWithoutValidation), you lose your original raw values.